TaskWithSmartPointers/MySharedPtr.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+#include <iostream>
+
+template <class T>
+class MySharedPtr
+{
+private:
+	T* m_ptr = nullptr;
+	long* m_referencesCounter = nullptr;
+	friend std::ostream& operator<<(std::ostream& os, MySharedPtr const& m);
+public:
+	T* get()
+	{
+		return m_ptr;
+	}
+
+	long use_count() const noexcept
+	{
+		if (m_referencesCounter != nullptr)
+		{
+			return *m_referencesCounter;
+		}
+		else
+			return 0;
+	}
+
+	bool unique() const noexcept
+	{
+		return (1 == use_count());
+	}
+
+	void reset(T* p)
+	{
+		if (m_referencesCounter != nullptr)
+		{
+			(*m_referencesCounter)--;
```

Comment:
I would write:
`(*m_referencesCounter) -= 1;`

As you know `(*m_referencesCounter)--;` will decrement the variable after the end of the scope. I'm not sure the condition at next line will be satisfied.

Have you checked the debugger hits that?

Reply:
Please check this for all cases.

Reply:
Fixed